backend/alert/alerts.py

```diff
@@ -48,46 +49,134 @@ def __init__(self, template, reg):
+        and None if notification was attempted to be sent but an error occurred.
+        """
+        if (
+            self.registration.user
+            and self.registration.user.profile
+            and self.registration.user.profile.push_notifications
+        ):
+            # Do not send text if push_notifications is enabled
+            return False
         if self.registration.user is not None and self.registration.user.profile.phone is not None:
             phone_number = self.registration.user.profile.phone
         elif self.registration.phone is not None:
             phone_number = self.registration.phone
         else:
             return False
 
-        return send_text(phone_number, self.text)
+        if close_notification:
+            if not self.close_text:
+                # This should be unreachable
+                return None
+            alert_text = self.close_text
+        else:
+            alert_text = self.text
+        return send_text(phone_number, alert_text)
+
+
+class PushNotification(Alert):
+    def __init__(self, reg):
+        super().__init__("alert/push_notif.txt", reg, close_template="alert/push_notif_close.txt")
+
+    def send_alert(self, close_notification=False):
+        """
+        Returns False if notification was not sent intentionally,
+        and None if notification was attempted to be sent but an error occurred.
+        """
+        if (
+            self.registration.user
+            and self.registration.user.profile
+            and self.registration.user.profile.push_notifications
+        ):
+            # Only send push notification if push_notifications is enabled
+            pennkey = self.registration.user.username
+            bearer_token = str(10101)
+            if close_notification:
+                if not self.close_text:
+                    # This should be unreachable
+                    return None
+                alert_title = "%s just closed." % self.registration.section.full_code
+                alert_body = self.close_text
+            else:
+                alert_title = "%s is now open!" % self.registration.section.full_code
+                alert_body = self.text
+            try:
+                response = requests.post(
+                    "https:/api.pennlabs.org/notifications/send/internal",
+                    data={"title": alert_title, "body": alert_body, "pennkey": pennkey,},
+                    headers={"Authorization": "Bearer %s" % bearer_token},
+                )
+                if response.status_code != 200:
+                    logger.exception(
+                        f"Push Notification {response.status_code} Response: {response.content}"
+                    )
+                    return None
+            except requests.exceptions.RequestException as e:
+                logger.exception(f"Push Notification Request Error: {e}")
+                return None
+            return True
+        return False
```

Comment:
Instead of returning `False` in line 182, shouldn't we be returning `None` since `self.registration.user` is `None` due to some error? (i.e. we do want to send a push notif, but there is an error preventing us from doing so)

Reply:
So this is actually behavior which is grandfathered in from the old PCA; self.registration.user is None on legacy registrations, so in that case it just wouldn't send any push notifications.  Not really an error state so I don't think returning None makes sense, but it's kind of an unreachable case now since there is no way of creating legacy registrations anymore.